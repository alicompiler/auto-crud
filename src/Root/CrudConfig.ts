import FieldConfig from "react-auto-form-core/dist/Field/FieldConfig";
import {PageConfig} from "../Page/PageConfig";

export interface CrudConfig {
    name: string;
    endpointRoot: string;
    routeRoot: string;
    pages?: PageConfig[];
    indexPage?: PageConfig;
    createPage?: PageConfig;
    updatePage?: PageConfig;
    deletePage?: PageConfig;
    detailsPage?: PageConfig;


    mainTitle?: string;
    fields: FieldConfig[];
    layout?: () => any;
    header?: () => any;
}




