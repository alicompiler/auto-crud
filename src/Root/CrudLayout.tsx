import * as React from 'react';
import {Component} from 'react';

class CrudLayout extends Component {
    render() {
        return (
            <div className={'__curd-layout'}>
                <h1>Auto - Crud</h1>
                <h1>Crud Layout</h1>
            </div>
        );
    }
}

export default CrudLayout;