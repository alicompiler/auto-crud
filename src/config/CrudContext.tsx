import * as React from "react";
import {CrudConfig} from "./Config";
import {DefaultConfigFixer} from "./ConfigFixer/ConfigFixer";
import {UIStateInitializer} from "./UIStateInitializer";

export interface ContextConfig<State = any> {
    config: CrudConfig;
    state: State;
    ui: UIState;
    updateState: (payload: Partial<State>) => void;
}

export interface UIState {
    pages: {
        [pageName: string]: any;
    };
    modals: {
        [modalName: string]: any;
    };
}

export function getInitialState(config: CrudConfig): Omit<ContextConfig, "updateState"> {
    const updatedConfig = new DefaultConfigFixer(config).fix();
    const uiState = new UIStateInitializer(updatedConfig).initialize();
    return {
        config: updatedConfig,
        ui: uiState,
        state: {
            pageTitle: ''
        }
    }
}

const defaultValue: ContextConfig = undefined as any;

export const CrudContext = React.createContext(defaultValue as any);

