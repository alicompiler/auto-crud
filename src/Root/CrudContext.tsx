import * as React from "react";
import {CrudConfig} from "./CrudConfig";
import {DefaultConfigInitializer} from "./ConfigInitializer/ConfigInitializer";
import {UIStateInitializer} from "./ConfigInitializer/UIStateInitializer";

export interface CrudContextValue<State = any> {
    config: CrudConfig;
    state: State;
    ui: UIState;
    updateState: (payload: Partial<State>) => void;
}

export interface UIState {
    pages: {
        [pageName: string]: any;
    };
    modals: {
        [modalName: string]: any;
    };
}

export function getInitialState(config: CrudConfig): Omit<CrudContextValue, "updateState"> {
    const updatedConfig = new DefaultConfigInitializer(config).initialize();
    const uiState = new UIStateInitializer(updatedConfig).initialize();
    return {
        config: updatedConfig,
        ui: uiState,
        state: {
            pageTitle: ''
        }
    }
}

const defaultValue: CrudContextValue = undefined as any;
export const CrudContext = React.createContext(defaultValue as any);

