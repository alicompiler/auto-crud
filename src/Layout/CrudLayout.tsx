import React, {Component} from 'react';
import {CrudContext, CrudContextValue} from "../Root/CrudContext";
import {Switch} from "react-router-dom";
import RoutesExtractor from "./RoutesExtractor";


class CrudLayout extends Component {

    private previousSwitch: any = null;

    render() {
        return (
            <div>
                {
                    this.getRouterSwitch()
                }
            </div>
        );
    }


    protected getRouterSwitch = () => {
        if (this.previousSwitch)
            return this.previousSwitch;
        const routesExtractor = new RoutesExtractor(this.getContext());
        this.previousSwitch = <Switch>
            {routesExtractor.getRoutes()}
        </Switch>;
        return this.previousSwitch;
    };

    protected getContext = (): CrudContextValue => this.context;

}

CrudLayout.contextType = CrudContext;

export default CrudLayout;