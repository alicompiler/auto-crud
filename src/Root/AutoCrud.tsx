import * as React from 'react';
import {CrudContext} from "./CrudContext";
import {DefaultConfigInitializer} from "./ConfigInitializer/ConfigInitializer";
import CrudRootHeader from "../components/CrudRootHeader/CrudRootHeader";
import CrudLayout from "../Layout/CrudLayout";
import {CrudConfig} from "./CrudConfig";


class AutoCrud extends React.Component<CrudConfig, any> {

    static defaultProps = {
        layout: () => <CrudLayout/>,
        header: () => <CrudRootHeader/>,
    };

    private readonly config: CrudConfig;

    constructor(props: CrudConfig) {
        super(props);
        this.state = {state: {}, uiState: {}};
        this.config = new DefaultConfigInitializer(this.props).initialize();
    }


    render() {
        const layout = this.props.layout!();
        const header = this.props.header!();

        return (
            <div className={'__curd-Root'}>
                <CrudContext.Provider value={this.getContextValue()}>
                    {header}
                    {layout}
                </CrudContext.Provider>
            </div>
        );
    }

    private getContextValue = () => {
        return {
            state: this.state,
            config: this.config,
            updateState: (payload: any) => this.setState(payload)
        };
    }
}

export default AutoCrud;