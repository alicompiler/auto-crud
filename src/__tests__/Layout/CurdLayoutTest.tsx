import * as React from "react";
import {configure, mount, shallow} from "enzyme";
import Adapter from "enzyme-adapter-react-16"
import {BrowserRouter} from "react-router-dom";
import CrudLayout from "../../Layout/CrudLayout";
import {CrudContext, getInitialState} from "../../Root/CrudContext";

configure({adapter: new Adapter()});

describe('CurdLayout', () => {
    it('should match snapshot', function () {
        const component = shallow(<BrowserRouter>
            <CrudContext.Provider value={getInitialState({name: '', endpointRoot: '', fields: []})}>
                <CrudLayout/>
            </CrudContext.Provider>
        </BrowserRouter>);

        expect(component).toMatchSnapshot();
    });
});