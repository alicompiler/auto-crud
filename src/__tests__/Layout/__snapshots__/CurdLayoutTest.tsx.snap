// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurdLayout should match snapshot 1`] = `ShallowWrapper {}`;
